### PR TITLE
Wieger tick cluster go

### PR DESCRIPTION
**Pull request by CorgiMan**
- spawn generic ringpop processes (binaries or via interpreter). e.g.
```
tickcluster 10 path/to/binary/testpop
tickcluster 10 -i node path/to/main.js
tickcluster 10 --interpreter node path/to/main.js
```
where tickcluster is an alias for `alias tickcluster='node ~/code/ringpop/scripts/tick-cluster.js'`.

- Adds grid mode view. Press z to activate

- Also removed a bug where tickcluster crashes if you request status (type 's') directly after starting the cluster.

@jwolski, @danielheller

### CODE REVIEW

**Comment by CorgiMan** [comment]
I made a sloppy mistake with git. Let me know if I need to squash for decency

File: scripts/tick-cluster.js [hunk 1]
```diff
@@ -146,6 +147,48 @@ function statsAll() {
     });
 }
 
+function grid() {
```
Comment (jwolski) [inline scripts/tick-cluster.js]: 'grid mode' should be separated out as its own PR. It's best to keep unrelated features separate so that it's easier to revert (if need be) and review.

File: scripts/tick-cluster.js [hunk 2]
```diff
@@ -429,7 +487,7 @@ function killProc(count) {
         var proc = procs[rand];
         if (proc.killed === null && proc.suspended === null) {
             logMsg(proc.port, color.green('pid ' + proc.pid) + color.red(' randomly selected for death'));
-            process.kill(proc.proc.pid, 'SIGTERM');
+            process.kill(proc.proc.pid, 'SIGKILL');
```
Comment (jwolski) [inline scripts/tick-cluster.js]: how does Go handle the SIGTERM?

File: scripts/tick-cluster.js [hunk 3]
```diff
@@ -472,6 +529,7 @@ function main() {
     stdin.resume();
     stdin.setEncoding('utf8');
     stdin.on('data', onData);
+    setInterval(grid, 50);
```
Comment (jwolski) [inline scripts/tick-cluster.js]: For completeness sake, you should `clearInterval()` upon quitting tick-cluster

File: scripts/tick-cluster.js [hunk 4]
```diff
@@ -527,10 +586,14 @@ function send(host, arg1, arg2, arg3, callback) {
 
 program
     .version(require('../package.json').version)
-    .arguments('<size>')
+    .option("-i, --interpreter [interpreter]")
```
Comment (jwolski) [inline scripts/tick-cluster.js]: If <program> and <size> are required params, where do we validate them? How about providing a sane default for "size" and error out if no program is specified?

File: scripts/tick-cluster.js (hunk 2 above)
Comment (CorgiMan) [inline scripts/tick-cluster.js]: I can't find it right now but it terminates the process. You can catch the signal with the os package to shut down gracefully.

We want to simulate a node failing here so I thought SIGKILL would be more appropriate.

File: scripts/tick-cluster.js (hunk 1 above)
Comment (CorgiMan) [inline scripts/tick-cluster.js]: I'll do that

File: scripts/tick-cluster.js [hunk 5]
```diff
@@ -541,8 +570,14 @@ program.on('--help', function onHelp() {
...
+
+if(!fs.existsSync(programPath)) {
+    console.log('Error: program ' + programPath + ' does not exist. Check path');
+    process.exit(1);
+}
+
+if(procsToStart === 'NaN') {
+    console.log('Error: number of processes to start is not an integer');
     process.exit(1);
 }
```
Comment (CorgiMan) [inline scripts/tick-cluster.js]: Checks for validity of programPath and Number of processes to start arguments

File: scripts/tick-cluster.js [hunk 6]
```diff
@@ -527,10 +547,19 @@ function send(host, arg1, arg2, arg3, callback) {
...
+    .action(function onAction(path, args, options) {
+        programInterpreter = options.interpreter;
+        if (options.N)
+            procsToStart = parseInt(options.N);
+        programPath = path;
+        if (programPath[0] !== '/') {
+            programPath = './' + programPath;
+        }
+        programArgs = args;
     });
```
Comment (CorgiMan) [inline scripts/tick-cluster.js]: size is an option now (default if 5)

program path is prepended with ./ in case it is a relative path. We do this because to launch a binary in the current directory we need to type "./program". Just typing "program" would not start a process. This has the added bonus that we don't look in the path for binaries which could lead to some unexpected errors.

File: scripts/tick-cluster.js [hunk 7]
```diff
@@ -26,8 +26,10 @@ var color = require('cli-color');
 var farmhash = require('farmhash').hash32;
 var generateHosts = require('./generate-hosts');
 var program = require('commander');
+var fs = require('fs');
 
-var hosts, procs, procsToStart, ringPool, localIP, toSuspend, toKill, tchannel; // defined later
+var programArgs, programInterpreter, programPath, procsToStart = 5;
+var hosts, procs, ringPool, localIP, toSuspend, toKill, tchannel; // defined later
```
Comment (CorgiMan) [inline scripts/tick-cluster.js]: default procsToStart = 5 (do we want to make this a seperate variable?)

**Comment by CorgiMan** [comment]
Update for usage:
size is now the '-n' option (default is 5)
e.g.:
```
tickcluster -n 10 path_to_binary 
tickcluster -n 10 -i node path_to_node_program
```

File: scripts/tick-cluster.js [hunk 8]
```diff
@@ -541,8 +567,14 @@ program.on('--help', function onHelp() {
...
 
-if (!procsToStart) {
-    console.log('Error: size is a required argument');
+
+if(!fs.existsSync(programPath)) {
+    console.log('Error: program ' + programPath + ' does not exist. Check path');
+    process.exit(1);
+}
+
+if(procsToStart === 'NaN') {
```
Comment (jwolski) [inline scripts/tick-cluster.js]: use `isNaN(procsToStart)`

File: scripts/tick-cluster.js [hunk 9]
```diff
@@ -541,8 +567,14 @@ program.on('--help', function onHelp() {
...
 
-if (!procsToStart) {
-    console.log('Error: size is a required argument');
+
+if(!fs.existsSync(programPath)) {
+    console.log('Error: program ' + programPath + ' does not exist. Check path');
+    process.exit(1);
+}
+
+if(isNaN(procsToStart)) {
```
Comment (CorgiMan) [inline scripts/tick-cluster.js]: isNan, in stead of, === 'NaN'

File: scripts/tick-cluster.js [hunk 10]
```diff
@@ -527,10 +544,19 @@ function send(host, arg1, arg2, arg3, callback) {
 
 program
     .version(require('../package.json').version)
-    .arguments('<size>')
+    .option('-i, --interpreter [interpreter]')
+    .option('-n [number of ringpop nodes]')
```
Comment (jwolski) [inline scripts/tick-cluster.js]: `.options('-n <nodes>')`

File: scripts/tick-cluster.js [hunk 11]
```diff
@@ -527,10 +544,19 @@ function send(host, arg1, arg2, arg3, callback) {
 
 program
     .version(require('../package.json').version)
-    .arguments('<size>')
+    .option('-i, --interpreter [interpreter]')
```
Comment (jwolski) [inline scripts/tick-cluster.js]: `<>` instead of `[]`

File: scripts/tick-cluster.js [hunk 12]
```diff
@@ -346,8 +347,24 @@ function findLocalIP() {
...
 function ClusterProc(port) {
-    var newProc = require('child_process').spawn('node', ['main.js', '--listen=' + localIP + ':' + port, '--hosts=./hosts.json']);
+    var newProc;
+
+    if (programInterpreter) {
+        newProc = require('child_process').spawn(programInterpreter, [programPath, '--listen=' + localIP + ':' + port, '--hosts=./hosts.json'].concat(programArgs));
+    } else {
+        newProc = require('child_process').spawn(programPath, ['--listen=' + localIP + ':' + port, '--hosts=./hosts.json'].concat(programArgs));
+    }
+    newProc.on('error', function(err) {
```
Comment (jwolski) [inline scripts/tick-cluster.js]: Print out the host:port of the node that died when an `error` is emitted

File: scripts/tick-cluster.js [hunk 13]
```diff
@@ -346,8 +347,24 @@ function findLocalIP() {
...
+        console.log(err + 'error: failed to spawn ' + programPath);
```
Comment (jwolski) [inline scripts/tick-cluster.js]: print out `err.message` instead of `err`

File: scripts/tick-cluster.js [hunk 14]
```diff
@@ -527,10 +544,19 @@ function send(host, arg1, arg2, arg3, callback) {
...
-    .arguments('<size>')
+    .option('-i, --interpreter [interpreter]')
+    .option('-n [number of ringpop nodes]')
+    .arguments('<program> [args...]')
     .description('tick-cluster is a tool that launches a ringpop cluster of arbitrary size')
-    .action(function onAction(size) {
-        procsToStart = +size;
+    .action(function onAction(path, args, options) {
+        programInterpreter = options.interpreter;
+        if (options.N)
```
Comment (jwolski) [inline scripts/tick-cluster.js]:
```
if (options.N) {

}
```
